framework/resource/retryresource/crud_resource_wrapper.go

```diff
@@ -0,0 +1,100 @@
+package retryresource
+
+import (
+	"context"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	originalframework "github.com/giantswarm/operatorkit/framework"
+	"github.com/giantswarm/operatorkit/framework/resource/internal"
+	"github.com/giantswarm/operatorkit/framework/resource/internal/framework"
+)
+
+// crudResourceWrapper is a specialized wrapper which wraps
+// *framework.CRUDResource.
+type crudResourceWrapper struct {
+	logger   micrologger.Logger
+	resource framework.Resource
+
+	backOff backoff.BackOff
+}
+
+func newCRUDResourceWrapper(config Config) (*crudResourceWrapper, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+	if config.Resource == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Resource must not be empty")
+	}
+
+	if config.BackOff == nil {
+		config.BackOff = backoff.NewExponentialBackOff()
+	}
+
+	// Wrap underlying resource Ops with retry logic. Underlying resource
+	// is a pointer so we can modify it in place.
+	{
+		underlying, ok := internal.Underlying(config.Resource).(*originalframework.CRUDResource)
+		if !ok {
+			return nil, microerror.Maskf(incompatibleUnderlyingResourceError, "expected %T", underlying)
+		}
+
+		c := crudResourceOpsWrapperConfig{
+			Logger: config.Logger,
+			Ops:    underlying.CRUDResourceOps,
+
+			BackOff: config.BackOff,
+		}
+
+		wrappedOps, err := newCRUDResourceWrapperOps(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		underlying.CRUDResourceOps = wrappedOps
+	}
+
+	r := &crudResourceWrapper{
+		logger: config.Logger.With(
+			"underlyingResource", internal.Underlying(config.Resource).Name(),
+		),
+		resource: config.Resource,
+
+		backOff: config.BackOff,
+	}
+
+	return r, nil
+}
+
+func (r *crudResourceWrapper) EnsureCreated(ctx context.Context, obj interface{}) error {
+	// Just call wrapped resource. Wrapped crudResourceOpsWrapper will do
+	// the retries.
+	err := r.resource.EnsureCreated(ctx, obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *crudResourceWrapper) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	// Just call wrapped resource. Wrapped crudResourceOpsWrapper will do
+	// the retries.
+	err := r.resource.EnsureDeleted(ctx, obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *crudResourceWrapper) Name() string {
+	return r.resource.Name()
+}
+
+// Wrapped implements internal.Wrapper interface.
+func (r *crudResourceWrapper) Wrapped() framework.Resource {
```

Comment:
The wording was "underlying" so far. Why is this now "wrapped"?

Reply:
Check the internal package again. Maybe wording isn't best, but this is internal. We can change that any time.

`internal.Wrapper.Wrapped` returns next wrapped. `internal.Underlying` drills down `Wrapped` till it finds the one that doesn't wrap anything. It's called several places in `resource` package so I thought it'd be nice to extract. Also now it's easily testable (I want to write additional tests for that). I can provide separate PR.

Reply:
I think it should be `Underlying`, because this is the wording so far. I do not see why we should change this now. When it is inappropriate, we should just change it everywhere.

Reply:
What do you mean by everywhere?

Reply:
I have no time for this now anymore. Do what you think is right.